### ISSUE TITLE
new cloudflare setup breaks purple plugin

### ISSUE THREAD

**Issue by penguin42** [matrix-org/purple-matrix#72]
purple doesn't like the new matrix.org setup with cloudflare in front of it.
(Invalid response from homeserver is the error I see).
I'm not sure, but I think it's the chunked transfer encoding that's upsetting it

**Comment by penguin42** [comment]
I've not figured this out yet.
It looks to me like the http-parser is calling the on_body callback for the 1st chunk, yet the README for the http-parser explictly says:
    'That is, a chunked encoding is decoded before being sent to the on_body callback.'
which would be great if it was.

(20:01:02) matrixprpl: unable to parse JSON: <data>:1:15002: Parse error: unexpected string constant "body", expected character `:'
(20:01:02) matrixprpl: DAGDEBUG: parser 1: ret_len=22715213 parsed1=15623 error=6 upgrade=0
(20:01:02) matrixprpl: DAGDEBUG: parser 2: ret_len=22715213 parsed1=0 error=6 upgrade=0
(20:01:02) matrixprpl: Error (the on_body callback failed) parsing HTTP response HTTP/1.1 200 OK^M

so in the 1st call to the http_parser_execute it's called the json parser on the 1st chunk before it's retuned.  The first chunk line says 3a9a which is 15002 so that about matches up (with some header to get to the 15623???).
I've got to leave it for this weekend, so if anyone has any better ideas please add them here.

**Comment by ScottAbbey** [comment]
> It looks to me like the http-parser is calling the on_body callback for the 1st chunk, yet the README for the http-parser explictly says:
'That is, a chunked encoding is decoded before being sent to the on_body callback.'
which would be great if it was.

It looks like the chunked encoding is handled, but the data is not necessarily all sent at once:

> http_data_cb does not return data chunks. It will be call arbitrarally
> many times for each string. E.G. you might get 10 callbacks for "on_url"
> each providing just a few characters more data.

And from https://github.com/nodejs/http-parser/issues/339#issuecomment-252693014:

> You will have your on_body callback called with chunks of the body, and you will know you've gotten them all due to the on_message_complete callback.

**Comment by ScottAbbey** [comment]
Since `on_header_field` and `on_header_value` were also `http_data_cb`, I tried using a `GString` to build up a single large string from each call to `_handle_body`, then moved the `json_parser_load_from_data` into a new `_handle_message_complete` for `http_parser_settings.on_message_complete`.

I built it and tried it out and it seems to work fine in a few minutes of testing.

I will not share the code because I do not know how to write C safely, but hopefully this gives you enough information to fix this in an appropriate manner.
